src/Datadog.Trace/Util/PropertyFetcher.cs

```diff
@@ -24,61 +23,73 @@ public PropertyFetcher(string propertyName)
         /// <summary>
         /// Gets the value of the property on the specified object.
         /// </summary>
+        /// <typeparam name="T">Type of the result.</typeparam>
         /// <param name="obj">The object that contains the property.</param>
         /// <returns>The value of the property on the specified object.</returns>
-        public object Fetch(object obj)
+        public T Fetch<T>(object obj)
         {
-            Type objType = obj.GetType();
+            return Fetch<T>(obj, obj.GetType());
+        }
 
+        /// <summary>
+        /// Gets the value of the property on the specified object.
+        /// </summary>
+        /// <typeparam name="T">Type of the result.</typeparam>
+        /// <param name="obj">The object that contains the property.</param>
+        /// <param name="objType">Type of the object</param>
+        /// <returns>The value of the property on the specified object.</returns>
+        public T Fetch<T>(object obj, Type objType)
+        {
             if (objType != _expectedType)
```

Comment:
I don't really understand why we do that. If we call Fetch with a different type, shouldn't we throw?
If it's done just for initialization, shouldn't we do that in the constructor?

Reply:
I'm not entirely sure why we did it this way where we reconstruct the fetcher if the type of the object is different (other than the fact that we ported this directly from dotnet/runtime). It seems like the only upside is that this allows us to resolve properties against multiple implementation types of the same abstract type, but I think we could accomplish this by getting the type of the generic parameter `T` and it would work for all method arguments.

Reply:
@lucaspimentel do you have any more insight on this?

Reply:
This file was copied verbatim from the dotnet repo (see [url](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs) at top of file). Unfortunately, that link is to `master` and the file has since changed. We also made [this one change](https://github.com/DataDog/dd-trace-dotnet/commit/0ee897ac88caaec7872807009515c489d13af2bb#diff-03ea678402f5d547eb5d427be0c0c583) which I think was to allow getting inherited properties (not declared in the given type).

My guess is that `DiagnosticSource` needed to handle changing types because they rely heavily on anonymous types. For example, a fetcher for a property called `Name` could be called once with
```
new { Name = "Foo" }
```
and then later with
```
new { Name = "Foo", Value = 1 }
```

... and each time the type changed, they generate a new delegate. I don't think we need to support this "feature" for our use cases.

Reply:
For reference, I think the version we used was from .NET Core 2.1: [source](https://github.com/dotnet/corefx/blob/release/2.1/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs)